File: go/referral_test.go
package afi

import (
	"math/big"
	"testing"

	"github.com/ethereum/go-ethereum/common"
)

func assertReferralSelector(t *testing.T, data []byte, method string) {
	t.Helper()
	if len(data) < 4 {
		t.Fatalf("%s: calldata too short: %d", method, len(data))
	}
	m, ok := referralRouterParsedABI.Methods[method]
	if !ok {
		t.Fatalf("%s: method not in ABI", method)
	}
	if string(data[:4]) != string(m.ID) {
		t.Errorf("%s: selector mismatch", method)
	}
}

var (
	refTokenIn  = common.HexToAddress("0x1111111111111111111111111111111111111111")
	refTokenOut = common.HexToAddress("0x2222222222222222222222222222222222222222")
	refReferrer = common.HexToAddress("0x3333333333333333333333333333333333333333")
	refUser     = common.HexToAddress("0x4444444444444444444444444444444444444444")
	refDelegate = common.HexToAddress("0x5555555555555555555555555555555555555555")
	refTo       = common.HexToAddress("0x6666666666666666666666666666666666666666")
)

func TestReferralRouterAddress(t *testing.T) {
	addr, net, err := ReferralRouterAddress(8453)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if net != NetworkBase {
		t.Errorf("expected NetworkBase, got %s", net)
	}
	if addr != ReferralRouterAddresses[NetworkBase] {
		t.Errorf("address mismatch: %s", addr)
	}
}

func TestReferralRouterAddress_UnknownChain(t *testing.T) {
	if _, _, err := ReferralRouterAddress(999999); err == nil {
		t.Error("expected error for unknown chain id, got nil")
	}
}

func TestEncodeSwapWithReferral(t *testing.T) {
	data, err := EncodeSwapWithReferral(refTokenIn, big.NewInt(1000), refTokenOut, big.NewInt(900), []byte{0x01, 0x02}, refReferrer, 5)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "swapWithReferral")

	// Round-trip the args to make sure they encode in the right order.
	args, err := referralRouterParsedABI.Methods["swapWithReferral"].Inputs.Unpack(data[4:])
	if err != nil {
		t.Fatalf("unpack: %v", err)
	}
	if args[0].(common.Address) != refTokenIn {
		t.Errorf("tokenIn mismatch: %v", args[0])
	}
	if args[1].(*big.Int).Cmp(big.NewInt(1000)) != 0 {
		t.Errorf("amountIn mismatch: %v", args[1])
	}
	if args[6].(uint16) != 5 {
		t.Errorf("referralBps mismatch: %v", args[6])
	}
}

func TestEncodeSwapWithReferral_ZeroAmount(t *testing.T) {
	if _, err := EncodeSwapWithReferral(refTokenIn, big.NewInt(0), refTokenOut, big.NewInt(0), nil, refReferrer, 5); err == nil {
		t.Error("expected error for zero amountIn, got nil")
	}
	if _, err := EncodeSwapWithReferral(refTokenIn, nil, refTokenOut, big.NewInt(0), nil, refReferrer, 5); err == nil {
		t.Error("expected error for nil amountIn, got nil")
	}
}

func TestEncodeSwapWithReferral_BpsTooHigh(t *testing.T) {
	if _, err := EncodeSwapWithReferral(refTokenIn, big.NewInt(1), refTokenOut, big.NewInt(0), nil, refReferrer, ReferralHardCapBps+1); err == nil {
		t.Error("expected error for referralBps > hard cap, got nil")
	}
}

func TestEncodeSwapWithReferralFor(t *testing.T) {
	data, err := EncodeSwapWithReferralFor(refUser, refTokenIn, big.NewInt(1000), refTokenOut, big.NewInt(900), nil, refReferrer, 10)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "swapWithReferralFor")
}

func TestEncodeSwapWithReferralFor_Errors(t *testing.T) {
	if _, err := EncodeSwapWithReferralFor(common.Address{}, refTokenIn, big.NewInt(1), refTokenOut, big.NewInt(0), nil, refReferrer, 5); err == nil {
		t.Error("expected error for zero user, got nil")
	}
	if _, err := EncodeSwapWithReferralFor(refUser, refTokenIn, big.NewInt(0), refTokenOut, big.NewInt(0), nil, refReferrer, 5); err == nil {
		t.Error("expected error for zero amountIn, got nil")
	}
	if _, err := EncodeSwapWithReferralFor(refUser, refTokenIn, big.NewInt(1), refTokenOut, big.NewInt(0), nil, refReferrer, ReferralHardCapBps+1); err == nil {
		t.Error("expected error for referralBps > hard cap, got nil")
	}
}

func TestEncodeReferralClaim(t *testing.T) {
	data, err := EncodeReferralClaim(refTokenOut, refTo)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(data) != 4+32+32 {
		t.Fatalf("claim: expected 68 bytes, got %d", len(data))
	}
	assertReferralSelector(t, data, "claim")
}

func TestEncodeReferralClaim_ZeroTo(t *testing.T) {
	if _, err := EncodeReferralClaim(refTokenOut, common.Address{}); err == nil {
		t.Error("expected error for zero to, got nil")
	}
}

func TestEncodeReferralClaimMany(t *testing.T) {
	data, err := EncodeReferralClaimMany([]common.Address{refTokenIn, refTokenOut}, refTo)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "claimMany")
}

func TestEncodeReferralClaimMany_Errors(t *testing.T) {
	if _, err := EncodeReferralClaimMany(nil, refTo); err == nil {
		t.Error("expected error for empty tokens, got nil")
	}
	if _, err := EncodeReferralClaimMany([]common.Address{refTokenIn}, common.Address{}); err == nil {
		t.Error("expected error for zero to, got nil")
	}
}

func TestEncodeSetDelegateAllowance(t *testing.T) {
	data, err := EncodeSetDelegateAllowance(refTokenIn, refDelegate, big.NewInt(500), 1893456000)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "setDelegateAllowance")
}

func TestEncodeSetDelegateAllowance_Errors(t *testing.T) {
	if _, err := EncodeSetDelegateAllowance(refTokenIn, common.Address{}, big.NewInt(1), 1); err == nil {
		t.Error("expected error for zero delegate, got nil")
	}
	if _, err := EncodeSetDelegateAllowance(refTokenIn, refDelegate, nil, 1); err == nil {
		t.Error("expected error for nil amount, got nil")
	}
	if _, err := EncodeSetDelegateAllowance(refTokenIn, refDelegate, big.NewInt(-1), 1); err == nil {
		t.Error("expected error for negative amount, got nil")
	}
	// amount exceeding uint208
	over208 := new(big.Int).Lsh(big.NewInt(1), 208)
	if _, err := EncodeSetDelegateAllowance(refTokenIn, refDelegate, over208, 1); err == nil {
		t.Error("expected error for amount > uint208, got nil")
	}
	// deadline exceeding uint48
	if _, err := EncodeSetDelegateAllowance(refTokenIn, refDelegate, big.NewInt(1), (1<<48)); err == nil {
		t.Error("expected error for deadline > uint48, got nil")
	}
}

func TestEncodeRevokeDelegate(t *testing.T) {
	data, err := EncodeRevokeDelegate(refTokenIn, refDelegate)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "revokeDelegate")
}

func TestEncodeRevokeDelegate_ZeroDelegate(t *testing.T) {
	if _, err := EncodeRevokeDelegate(refTokenIn, common.Address{}); err == nil {
		t.Error("expected error for zero delegate, got nil")
	}
}

func TestEncodeReferralPauseUnpause(t *testing.T) {
	d1, err := EncodeReferralPause()
	if err != nil || len(d1) != 4 {
		t.Fatalf("pause: err=%v len=%d", err, len(d1))
	}
	assertReferralSelector(t, d1, "pause")

	d2, err := EncodeReferralUnpause()
	if err != nil || len(d2) != 4 {
		t.Fatalf("unpause: err=%v len=%d", err, len(d2))
	}
	assertReferralSelector(t, d2, "unpause")
}

func TestEncodeReferralSetMaxReferralBps(t *testing.T) {
	data, err := EncodeReferralSetMaxReferralBps(ReferralHardCapBps)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "setMaxReferralBps")

	if _, err := EncodeReferralSetMaxReferralBps(ReferralHardCapBps + 1); err == nil {
		t.Error("expected error for bps > hard cap, got nil")
	}
}

func TestEncodeReferralRescueTokens(t *testing.T) {
	data, err := EncodeReferralRescueTokens(refTokenOut, refTo)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertReferralSelector(t, data, "rescueTokens")

	if _, err := EncodeReferralRescueTokens(refTokenOut, common.Address{}); err == nil {
		t.Error("expected error for zero to, got nil")
	}
}

File: nodejs/src/__tests__/referral.test.ts
import { describe, expect, it } from "vitest"
import { decodeFunctionData, toFunctionSelector } from "viem"
import { AFI_REFERRAL_ROUTER_ABI } from "../constants.js"
import {
  REFERRAL_HARD_CAP_BPS,
  referralRouterAddress,
  encodeSwapWithReferral,
  encodeSwapWithReferralFor,
  encodeReferralClaim,
  encodeReferralClaimMany,
  encodeSetDelegateAllowance,
  encodeRevokeDelegate,
  encodeReferralPause,
  encodeReferralUnpause,
  encodeReferralSetMaxReferralBps,
  encodeReferralRescueTokens,
} from "../referral.js"

const TOKEN_IN  = "0x1111111111111111111111111111111111111111" as const
const TOKEN_OUT = "0x2222222222222222222222222222222222222222" as const
const REFERRER  = "0x3333333333333333333333333333333333333333" as const
const USER      = "0x4444444444444444444444444444444444444444" as const
const DELEGATE  = "0x5555555555555555555555555555555555555555" as const
const TO        = "0x6666666666666666666666666666666666666666" as const
const ZERO      = "0x0000000000000000000000000000000000000000" as const

function assertCall(data: string, fnName: string, signature: string) {
  expect(data).toMatch(/^0x[0-9a-f]+$/i)
  expect(data.startsWith(toFunctionSelector(signature))).toBe(true)
  const decoded = decodeFunctionData({ abi: AFI_REFERRAL_ROUTER_ABI, data: data as `0x${string}` })
  expect(decoded.functionName).toBe(fnName)
  return decoded
}

describe("referralRouterAddress", () => {
  it("resolves deployed chains", () => {
    expect(referralRouterAddress(8453)).toBe("0x2dC7a3990618baa91c450521004F14A334BF47c6")
    expect(referralRouterAddress(1)).toBe("0x47E7cE4237130F02202e081Efa1Fd338F23Ead77")
  })
  it("throws for an undeployed chain", () => {
    expect(() => referralRouterAddress(999999)).toThrow()
  })
})

describe("encodeSwapWithReferral", () => {
  it("encodes swapWithReferral(...)", () => {
    const data = encodeSwapWithReferral({
      tokenIn: TOKEN_IN, amountIn: 1000n, tokenOut: TOKEN_OUT, minOut: 900n,
      params: "0x0102", referrer: REFERRER, referralBps: 5,
    })
    const decoded = assertCall(data, "swapWithReferral",
      "swapWithReferral(address,uint256,address,uint256,bytes,address,uint16)")
    const args = decoded.args as readonly unknown[]
    expect(args[0]).toBe(TOKEN_IN)
    expect(args[1]).toBe(1000n)
    expect(args[6]).toBe(5)
  })
  it("rejects zero amountIn", () => {
    expect(() => encodeSwapWithReferral({
      tokenIn: TOKEN_IN, amountIn: 0n, tokenOut: TOKEN_OUT, minOut: 0n,
      params: "0x", referrer: REFERRER, referralBps: 5,
    })).toThrow()
  })
  it("rejects referralBps above the hard cap", () => {
    expect(() => encodeSwapWithReferral({
      tokenIn: TOKEN_IN, amountIn: 1n, tokenOut: TOKEN_OUT, minOut: 0n,
      params: "0x", referrer: REFERRER, referralBps: REFERRAL_HARD_CAP_BPS + 1,
    })).toThrow()
  })
})

describe("encodeSwapWithReferralFor", () => {
  it("encodes swapWithReferralFor(...)", () => {
    const data = encodeSwapWithReferralFor({
      user: USER, tokenIn: TOKEN_IN, amountIn: 1000n, tokenOut: TOKEN_OUT, minOut: 900n,
      params: "0x", referrer: REFERRER, referralBps: 10,
    })
    assertCall(data, "swapWithReferralFor",
      "swapWithReferralFor(address,address,uint256,address,uint256,bytes,address,uint16)")
  })
  it("rejects zero user / zero amount / bps too high", () => {
    const base = { tokenIn: TOKEN_IN, tokenOut: TOKEN_OUT, minOut: 0n, params: "0x" as const, referrer: REFERRER }
    expect(() => encodeSwapWithReferralFor({ ...base, user: ZERO, amountIn: 1n, referralBps: 5 })).toThrow()
    expect(() => encodeSwapWithReferralFor({ ...base, user: USER, amountIn: 0n, referralBps: 5 })).toThrow()
    expect(() => encodeSwapWithReferralFor({ ...base, user: USER, amountIn: 1n, referralBps: REFERRAL_HARD_CAP_BPS + 1 })).toThrow()
  })
})

describe("encodeReferralClaim", () => {
  it("encodes claim(token,to)", () => {
    const decoded = assertCall(encodeReferralClaim(TOKEN_OUT, TO), "claim", "claim(address,address)")
    const args = decoded.args as readonly unknown[]
    expect(args[0]).toBe(TOKEN_OUT)
    expect(args[1]).toBe(TO)
  })
  it("rejects zero to", () => {
    expect(() => encodeReferralClaim(TOKEN_OUT, ZERO)).toThrow()
  })
})

describe("encodeReferralClaimMany", () => {
  it("encodes claimMany(tokens,to)", () => {
    assertCall(encodeReferralClaimMany([TOKEN_IN, TOKEN_OUT], TO), "claimMany", "claimMany(address[],address)")
  })
  it("rejects empty tokens and zero to", () => {
    expect(() => encodeReferralClaimMany([], TO)).toThrow()
    expect(() => encodeReferralClaimMany([TOKEN_IN], ZERO)).toThrow()
  })
})

describe("encodeSetDelegateAllowance", () => {
  it("encodes setDelegateAllowance(token,delegate,uint208,uint48)", () => {
    const decoded = assertCall(
      encodeSetDelegateAllowance(TOKEN_IN, DELEGATE, 500n, 1893456000),
      "setDelegateAllowance",
      "setDelegateAllowance(address,address,uint208,uint48)",
    )
    const args = decoded.args as readonly unknown[]
    expect(args[2]).toBe(500n)
    expect(args[3]).toBe(1893456000)
  })
  it("rejects zero delegate / over-uint208 amount / over-uint48 deadline", () => {
    expect(() => encodeSetDelegateAllowance(TOKEN_IN, ZERO, 1n, 1)).toThrow()
    expect(() => encodeSetDelegateAllowance(TOKEN_IN, DELEGATE, 1n << 208n, 1)).toThrow()
    expect(() => encodeSetDelegateAllowance(TOKEN_IN, DELEGATE, 1n, 2 ** 48)).toThrow()
  })
})

describe("encodeRevokeDelegate", () => {
  it("encodes revokeDelegate(token,delegate)", () => {
    assertCall(encodeRevokeDelegate(TOKEN_IN, DELEGATE), "revokeDelegate", "revokeDelegate(address,address)")
  })
  it("rejects zero delegate", () => {
    expect(() => encodeRevokeDelegate(TOKEN_IN, ZERO)).toThrow()
  })
})

describe("owner-only controls", () => {
  it("encodes pause() / unpause()", () => {
    assertCall(encodeReferralPause(), "pause", "pause()")
    assertCall(encodeReferralUnpause(), "unpause", "unpause()")
  })
  it("encodes setMaxReferralBps(uint16) and rejects above cap", () => {
    assertCall(encodeReferralSetMaxReferralBps(REFERRAL_HARD_CAP_BPS), "setMaxReferralBps", "setMaxReferralBps(uint16)")
    expect(() => encodeReferralSetMaxReferralBps(REFERRAL_HARD_CAP_BPS + 1)).toThrow()
  })
  it("encodes rescueTokens(token,to) and rejects zero to", () => {
    assertCall(encodeReferralRescueTokens(TOKEN_OUT, TO), "rescueTokens", "rescueTokens(address,address)")
    expect(() => encodeReferralRescueTokens(TOKEN_OUT, ZERO)).toThrow()
  })
})
